lib/mix/lib/mix/dep.ex

```diff
@@ -132,6 +132,17 @@ defmodule Mix.Dep do
     seen = populate_seen(MapSet.new(), [app])
     children = get_deps(deps, tl(Enum.uniq(get_children(deps, seen, [app]))))
 
+    if bottom.project()[:apps_path] do
```

Comment:
So we can't call `bottom.project()` directly because users may have code in their `def project` that assumes we are running it in the project root directory. 

After taking a quick look, I think the best place to implement this check is here:

https://github.com/elixir-lang/elixir/blob/master/lib/mix/lib/mix/dep/loader.ex#L301

You are already inside the dependency, and you have both the cached project options (by calling `Mix.Project.config()`) and the dependency. Then you can compare `Mix.Project.config()[:app]` with `dep.app` and raise if they are different.

We will also need a test, you can look at `mix/umbrella_test.exs` for an example. But please don't create a new fixture, you can use one of the existing features and change the `mix.exs` of an umbrella child in the middle of the test by calling `File.write!`.

Thanks for looking into this, it is really appreciated!

Reply:
I really appreciate your guidance. Right now I'm heading to Elixir Brasil conf, and I'll work on that on next week :)